### ISSUE TITLE
[HOLD for payment 2023-03-09] [$1000] Android app does not focus on date of birth field when 'fix the errors' link is clicked

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15290]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app in android
2. Open settings
3. Open Profile
4. Open Personal details
5. Open date of birth
6. Write or select any wrong date eg: 02/02/2024
7. Click on save
8. Click on fix the errors

## Expected Result:
App should focus on date of birth field and open the calender when 'fix the errors' is clicked like it does on mWeb android chrome

## Actual Result:
App does nothing on 'fix the errors' click on date of birth page (same issue on connect bank account step 3 and step 2 i.e. issue exists on all the pages with calender selection on android app)

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] iOS / native

**Version Number:** 1.2.74-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/219977038-dbb657f0-a530-493f-9dd9-0d1096039f1b.mp4

https://user-images.githubusercontent.com/43996225/219977054-d18c34b8-95fb-467c-8955-d496b6cdcd82.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1676817426854809

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01996698b9e9290a5b</li>
        <li>Upwork Job ID: 1627953353378811904</li>
        <li>Last Price Increase: 2023-02-21</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sakluger (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by mountiny** [comment]
Callstack can take this one

**Comment by narefyev91** [comment]
Hello, I'm Nicolay from Callstack and I'm interested in taking and analysing this issue to work on a fix for it.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01996698b9e9290a5b

**Comment by MelvinBot** [comment]
Current assignee @sakluger is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @Santhosh-Sellavel (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @alex-mechler (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by MelvinBot** [comment]
📣 @narefyev91 You have been assigned to this job by @Beamanator!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by suryaneelankar** [comment]
On Clicking fix the errors we need to show the modal calendar again using useState with which it will get the expected result

**Comment by Santhosh-Sellavel** [comment]
Let me know when PR is ready @narefyev91, thanks!

**Comment by narefyev91** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Native apps **don't open the datepicker** when clicking on **Fix error button**

### What is the root cause of that problem?
The main issue here - combination of trying to _focus input_ (clicking on fix error button) and _editable={false}_ for input field

### What changes do you think we should make in order to solve the problem?
Case is special because global Form error handling (src/components/Form.js) doesn't even know that any field maybe not editable (focusable) and to be honest it should not know about it - error logic should work for all fields.
Date of birthday field is not editable - and this code will not work
```
  if (focusInput.focus && typeof focusInput.focus === 'function') {
      focusInput.focus();
  }
```
BTW we can implement custom function to fire on focus event. 
Here we see that we need to have `custom ref` instead of `normal ref` for input. If it will be functional components - we will use [useImperativeHandle](https://beta.reactjs.org/reference/react/useImperativeHandle) - which specified to apply custom handlers for any default ref functions.
We can't use this hook here - in that case we need to modify TextInput ref to make it working based on our needs and not breaking any existing functionality
Solution will be to add showPicker method for focus event
```
  ref={(el) => {
      if (!_.isFunction(this.props.innerRef)) {
          return;
      }
      if (el && el.focus && typeof el.focus === 'function') {
          let inputRef = {...el};
          inputRef = {...inputRef, focus: this.showPicker};
          this.props.innerRef(inputRef);
          return;
      }

      this.props.innerRef(el);
  }}
```

### What alternative solutions did you explore?
Alternative solution was playing around with making Input editable - but it will show some weird behaviour  - like jumping keyboard/datepicker

cc @Santhosh-Sellavel @mountiny

**Comment by mountiny** [comment]
@alex-mechler is the engineer assigned here so I will leave this proposal review to them 🙌

**Comment by alex-mechler** [comment]
That proposal looks good to me @narefyev91. @Santhosh-Sellavel do you have anything to add about it?

**Comment by Santhosh-Sellavel** [comment]
All good for now, I will follow up on the PR thanks!

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.77-4 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15473


If no regressions arise, payment will be issued on 2023-03-09. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@Santhosh-Sellavel / @alex-mechler] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@Santhosh-Sellavel / @alex-mechler] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@Santhosh-Sellavel / @alex-mechler] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@narefyev91] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@sakluger] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by sakluger** [comment]
@Santhosh-Sellavel I sent you an offer through Upwork.

Hi @narefyev91 - could you please apply to the Upwork job (https://www.upwork.com/jobs/~01996698b9e9290a5b) or comment with your Upwork profile links so that I can send you an offer via Upwork?

**Comment by TomaszFrackowiak** [comment]
@sakluger - Hi, @narefyev91 is from Callstack, so you don't need to send an offer via Upwork

**Comment by MelvinBot** [comment]
📣 @TomaszFrackowiak! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by Santhosh-Sellavel** [comment]
Accepted now @sakluger Sorry for the delay!

**Comment by Santhosh-Sellavel** [comment]
@sakluger @alex-mechler 
Does this qualify for an efficiency bonus or not?

**Comment by alex-mechler** [comment]
Proposal was accepted at 2/23 9:46am PT. PR was merged 2/27 at 8:59am PT. The 25th and 26th were Saturday and Sunday, so this falls in the 3 business day window. I think this qualifies for the bonus @sakluger

**Comment by alex-mechler** [comment]
On hold for payment tomorrow melvin

**Comment by sakluger** [comment]
Paid out everyone and closed the Upwork post!

@narefyev91 could you please propose if we need regression testing steps, then check off the checkbox in the BugZero checklist above? Thanks!

**Comment by narefyev91** [comment]
> Paid out everyone and closed the Upwork post!
> 
> @narefyev91 could you please propose if we need regression testing steps, then check off the checkbox in the BugZero checklist above? Thanks!

@sakluger in my view no regression tests needed here

**Comment by sakluger** [comment]
Sounds good @narefyev91! On future issues, feel free to check off the checkbox in the BugZero checklist and leave your proposal in a comment (even if that proposal is "No regression steps".

**Comment by Santhosh-Sellavel** [comment]
There is nothing to do here,

1. Date field is the only field here and highlight with the error
2. We have the calendar always open, so the user can just pick the validate now.

**Comment by Beamanator** [comment]
(I think wrong Alex was pinged here? Leaving for you @alex-mechler )

**Comment by alex-mechler** [comment]
Is this only happening on that specific input @kbecciv? If so, that input is non-focusable, since the calendar is always open, and thus not an issue for the reasons that @Santhosh-Sellavel mentioned. 

If this is happening across every form in the app, then its an issue and we should open a new issue for it
